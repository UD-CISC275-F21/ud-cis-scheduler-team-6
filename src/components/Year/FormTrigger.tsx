import React, {useRef} from "react";
import Col from "react-bootstrap/Col";
import Popover from "react-bootstrap/Popover";
import PopoverContent from "react-bootstrap/PopoverContent";
import Overlay from "react-bootstrap/Overlay";


interface FormTriggerProps{
    currentForm: string | null,
    setForm: (newForm: string | null) => void,
    YearUuid: string,
    children: JSX.Element
}

export default function FormTrigger(props: FormTriggerProps): JSX.Element{
    const overlayButton = useRef(null);
    return (
        <Col data-testid="form-trigger">
            <button
                data-testid="open-semester-form"
                className="trigger"
                ref={overlayButton}
                onClick={() => {
                    props.setForm(
                        props.currentForm === props.YearUuid
                            ? null
                            : props.YearUuid
                    );
                }}
            >
                +
            </button>
            <Overlay
                target={overlayButton}
                placement="right-end"
                show={props.currentForm === props.YearUuid}
                onHide={() => {
                    props.setForm(null);
                }}
                rootClose={true}
                transition={false}
            >
                <Popover id="popover-basic" data-testid="popover">
                    <PopoverContent>
                        <div
                            data-testid={"semester-form"}
                        >
                            {props.children}
                        </div>
                    </PopoverContent>
                </Popover>
            </Overlay>
        </Col>
    );
}