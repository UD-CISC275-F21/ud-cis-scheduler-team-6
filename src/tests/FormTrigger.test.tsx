import React from "react";
import {screen, render} from "@testing-library/react";
import FormTrigger from "../components/Year/FormTrigger";
import {v4 as uuid} from "uuid";

describe(FormTrigger,() => {
    const doNothingWithString = jest.fn<void, [string | null]>();
    it("Should render the button to open the popup, but not the popup itself if the currentForm is null.", async () => {
        render(<FormTrigger
            currentForm={null}
            setForm={doNothingWithString}
            YearUuid={uuid()}
        ><div>I should be hidden</div>
        </FormTrigger>);
        expect(screen.getByTestId("open-semester-form")).toBeInTheDocument();
        expect(screen.queryByText("I should be hidden")).not.toBeInTheDocument();
    });
    it("Should be the same no matter what currentForm is as long as it is not YearUuid", async () => {
        const {rerender} = render(<FormTrigger
            currentForm={null}
            setForm={doNothingWithString}
            YearUuid={uuid()}
        ><div>I should be hidden</div>
        </FormTrigger>);

        const html = screen.getByTestId("form-trigger").innerHTML;
        rerender(<FormTrigger
            currentForm={"x"}
            setForm={doNothingWithString}
            YearUuid={uuid()}
        ><div>I should be hidden</div>
        </FormTrigger>
        );
        expect(screen.getByTestId("form-trigger")).toContainHTML(html);
    });
    it("Should open its child in a popover if currentForm matches YearUuid", async () => {
        const YearUuid = uuid();
        render(<FormTrigger
            currentForm={YearUuid}
            setForm={doNothingWithString}
            YearUuid={YearUuid}
        ><div>I should not be hidden</div>
        </FormTrigger>
        );

        const popover = await screen.findByTestId("popover");
        expect(popover).toContainElement(screen.getByText("I should not be hidden"));
    });

    it("Should call setForm with YearUuid as the argument if the open-semester-form button is clicked", async () => {
        const YearUuid = uuid();
        const setFormSpy = jest.fn<void, [string | null]>();
        render(<FormTrigger
            currentForm={null}
            setForm={setFormSpy}
            YearUuid={YearUuid}
        ><div>I should be hidden</div>
        </FormTrigger>
        );
        expect(setFormSpy).not.toHaveBeenCalled();
        screen.getByTestId("open-semester-form").click();
        expect(setFormSpy).toHaveBeenCalled();
        expect(setFormSpy).toHaveBeenLastCalledWith(YearUuid);
    });
    it("Should call setForm with null as the argument if the open-semester-form button is pressed while the form is open.", async () => {
        const setFormSpy = jest.fn<void, [string | null]>();
        const YearUuid = uuid();
        render(<FormTrigger
            currentForm={YearUuid}
            setForm={setFormSpy}
            YearUuid={YearUuid}
        ><div>I should be hidden</div>
        </FormTrigger>
        );
        await screen.findByTestId("popover");
        expect(setFormSpy).not.toHaveBeenCalled();
        screen.getByTestId("open-semester-form").click();
        expect(setFormSpy).toHaveBeenCalled();
        expect(setFormSpy).toHaveBeenLastCalledWith(null);
    });
});