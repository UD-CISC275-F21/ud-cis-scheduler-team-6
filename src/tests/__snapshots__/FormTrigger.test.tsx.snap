// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTrigger Should only render a button if currentForm is null 1`] = `
<div
  className="col"
  data-testid="form-trigger"
>
  <button
    className="trigger"
    data-testid="open-semester-form"
    onClick={[Function]}
  >
    +
  </button>
</div>
`;
